crucible-symio/src/What4/CachedArray.hs

```diff
@@ -0,0 +1,837 @@
+-----------------------------------------------------------------------
+-- |
+-- Module           : What4.CachedArray
+-- Description      : What4 array storage with a concrete backing supporting symbolic indexes
+-- Copyright        : (c) Galois, Inc 2020
+-- License          : BSD3
+-- Maintainer       : Daniel Matichuk <dmatichuk@galois.com>
+-- Stability        : provisional
+------------------------------------------------------------------------
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE FlexibleInstances, FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- must come after TypeFamilies, see also https://gitlab.haskell.org/ghc/ghc/issues/18006
+{-# LANGUAGE NoMonoLocalBinds #-}
+
+module What4.CachedArray
```

Comment:
It would be nice to have some overall sense of what is going on in this module; it's not obvious just from the exposed API.

I guess it eventually wants to find it's way down into What4... is it ready for that?

Reply:
@danmatichuk Can you comment on what the CachedArray is trying to accomplish and whether or not it might be more generic than this memory model? I mostly left that part alone since I'm honestly not quite sure. It seemed like a deferred calculation of the contents of array fragments, but that is as far as I got...

Reply:
It's a general storage structure for What4 that is intended to support arrays which have data at a mix of both concrete and symbolic indexes. Logically it's just a multi-dimensional array, but supports reading and writing out "chunks" to the underlying store (i.e where a chunk spans continuous locations in the first index of the array). When indexing into an N-dimensional array, we take the abstract domain of N expressions and use it as an index into an N-level nested interval map. 

For the filesystem, this allows us to easily define a file that is fully concrete initially (i.e. chunk with a domain spanning the file contents), while efficiently supporting writes on top of this file at (potentially) symbolic locations.

There is probably an argument for merging this idea with the existing ConcreteArray in What4.

Reply:
OK, I think those would be helpful comments to add to the module documentation.  I can see this being a useful addition to What4, but I don't know that we should hold up this PR for that.

Reply:
Okay, I added this documentation and a bit more that I think is probably correct